woocommerce/woocommerce
# Add php unit watcher and update unit test docs

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





When writing unit tests, it's often necessary to run them multiple times to ensure they pass correctly.

In JavaScript, Jest offers a `--watch` mode, which monitors file changes and automatically runs tests. This feature proves invaluable during JS development and test writing.

This PR introduces the addition of the [phpunit-watcher ](https://github.com/spatie/phpunit-watcher)package to the `composer.json` file and commands to the `package.json` file. These additions enable tests to rerun automatically when files are modified. The [woocommerce-payments](https://github.com/Automattic/woocommerce-payments) repository has successfully utilized this package for some time.

Furthermore, this PR updates the unit test documentation to mandate the usage of the watcher and provides instructions for setting up the test environment using wp-env.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Review the changes
2. Make sure you can run the tests by following the instructions in the updated docs https://github.com/woocommerce/woocommerce/blob/dev/add-php-unit-watcher-and-unit-test-docs/plugins/woocommerce/tests/README.md

Please note that, unlike Jest, `phpunit-watcher` is unable to automatically only rerun the tests that are affected by the changes. It reruns all the tests. This is a limitation of the package. I also noticed that it's a bit slow when typing in the terminal in interactive mode but it still works fine when you press enter.


https://github.com/woocommerce/woocommerce/assets/4344253/344bfc64-4f14-4470-9dba-9a8507da92db




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Add php unit watcher and update unit test docs 

#### Comment 

</details>
